### PR TITLE
Refactor hash function templates and `hash_combine`

### PR DESCRIPTION
**Pull request by bdice**
This PR refactors a few pieces of libcudf's hash functions:
- Define the utility function `hash_combine` only once (with 32/64 bit overloads), rather than several times in the codebase
- ~Remove class template parameter from `MurmurHash3_32` and related classes. This template parameter was redundant. We already use a template for the argument of the `compute` method, which is called by `operator()`, so I put the template parameter on `operator()` instead of the whole class. I think this removal of the template parameter could be considered API-breaking so I added the `breaking` label.~ I retracted this change after conversation with @jrhemstad. I'll look into a different way to do this soon, using a dispatch-to-invoke approach as in #8217.

This addresses part of issue #10081. I have a few more things I'd like to try, but this felt like a nicely-scoped PR so I stopped here for the moment.

I benchmarked the code before and after making these changes and saw a small but consistent decrease in runtime.

The benchmarks in `HashBenchmark/{HASH_MURMUR3,HASH_SERIAL_MURMUR3,HASH_SPARK_MURMUR3}_{nulls,no_nulls}/*` all decreased or saw no change in runtime, with a geometric mean of 2.87% less time.

The benchmarks in `Hashing/hash_partition/*` all decreased or saw no change in runtime, with a geometric mean of 2.37% less time.

For both sets of benchmarks, the largest data sizes saw more significant decreases in runtime, with a best-improvement of 7.38% less time in `HashBenchmark/HASH_MURMUR3_nulls/16777216` (similar for other large data sizes) and a best-improvement of 10.54% less time in `Hashing/hash_partition/1048576/256/64` (similar for other large data sizes).

### CODE REVIEW

File: cpp/include/cudf/detail/hashing.hpp [hunk 1]
```diff
@@ -66,15 +87,15 @@ std::unique_ptr<column> serial_murmur_hash3_32(
  * Adapted from Boost hash_combine function, modified for 64-bit
  * https://www.boost.org/doc/libs/1_35_0/doc/html/boost/hash_combine_id241013.html
  *
- * @param lhs The first hashed value
- * @param rhs The second hashed value
- * @return Combined hash value
+ * @param lhs The first hash value.
+ * @param rhs The second hash value.
+ * @return Combined hash value.
```
Comment (vuule) [inline cpp/include/cudf/detail/hashing.hpp:92]: The smallest nitpick possible - I thought we don't end @param/@return comments with periods.

Comment (bdice) [reply cpp/include/cudf/detail/hashing.hpp:92]: I have been making this mistake a lot, if that's the case. I'm planning to give a Better Code talk about writing Doxygen docstrings in the future, so I'd appreciate a reference on this if you have one.
```suggestion
 * @param lhs The first hash value
 * @param rhs The second hash value
 * @return Combined hash value
```

Comment (vuule) [reply cpp/include/cudf/detail/hashing.hpp:92]: My comment is based on https://github.com/rapidsai/cudf/blob/branch-22.04/cpp/docs/DOCUMENTATION.md

Comment (bdice) [reply cpp/include/cudf/detail/hashing.hpp]: Thank you! It's almost too easy to point to the developer docs. 😅

**Review by vuule — APPROVED** [review]
Thanks for the detailed description, made it very easy to review the code :)
Looks good 👍

File: cpp/include/cudf/detail/utilities/hash_functions.cuh [hunk 2]
```diff
@@ -83,16 +83,15 @@ void __device__ inline uint32ToLowercaseHexString(uint32_t num, char* destinatio
 // algorithms are optimized for their respective platforms. You can still
 // compile and run any of them on any platform, but your performance with the
 // non-native version will be less than optimal.
-template <typename Key>
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:86]: The precedent in C++ hash function objects is that the type itself is a template and callable operator. 

https://en.cppreference.com/w/cpp/utility/hash

Comment (bdice) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:86]: @jrhemstad I wasn't familiar with that, thanks for the reference. Do you think we need to retain that class template behavior, or is this comment just informational to explain the previous design? In practice, we don't have any constructor behavior that varies by key type. It seems like a hash functor with type specializations of `operator()` is more in line with the rest of libcudf's design, unless you think that matching `std::hash` is crucial here.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh [hunk 3]
```diff
@@ -378,52 +346,72 @@ struct SparkMurmurHash3_32 {
...
+  return this->compute_floating_point(key);
+}
+
+template <>
+hash_value_type __device__ inline SparkMurmurHash3_32::operator()(
+  cudf::string_view const& key) const
+{
+  auto const data = reinterpret_cast<std::byte const*>(key.data());
+  auto const len  = key.size_bytes();
+  return this->compute_bytes(data, len);
```
Comment (codereport) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:396]: :heart:

**Review by codereport — APPROVED** [review]
LGreatTM :+1:

File: cpp/include/cudf/detail/utilities/hash_functions.cuh (hunk 2 above)
Comment (jrhemstad) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:86]: I'd prefer to keep the class template. 

> It seems like a hash functor with type specializations of operator() is more in line with the rest of libcudf's design

Class templates are actually more in line with "dispatch to invoke" pattern I've advocated for in the past. See: https://github.com/rapidsai/cudf/pull/8217/files and the Better Code talk I gave about this pattern.

**Comment by bdice** [comment]
rerun tests

**Comment by bdice** [comment]
I wanted to check the benchmarks again before merging. Commit 65edea2 shows the same performance as `branch-22.04` on average, but benchmarks of small data sizes are about 2% slower in 65edea2 than `branch-22.04` and benchmarks of large data sizes are about 6% faster in 65edea2 than `branch-22.04`. Since the performance has improved for large data sizes and shows only a small slowdown for small data sizes, I will merge this.

**Comment by bdice** [comment]
@gpucibot merge

**Pull request merged**